storage/innobase/srv/srv0srv.cc

```diff
@@ -2761,6 +2770,199 @@ srv_enable_temp_encryption_if_set()
 	}
 }
 
+/** Enable the undo log encryption if it is set.
+It will try to enable the undo log encryption and write the metadata to
+undo log file header, if innodb_undo_log_encrypt is ON. */
+static
+void
+srv_enable_undo_encryption_if_set()
+{
+	fil_space_t*	space;
+	const char*	cant_set_undo_tablespace = "Can't set undo tablespace";
+	const char*	to_be_encrypted = " to be encrypted";
+
+	if (srv_shutdown_state != SRV_SHUTDOWN_NONE) {
+		return;
+	}
+
+	/* Check if encryption for undo log is enabled or not. If it's
+	enabled, we will store the encryption metadata to the space header
+	and start to encrypt the undo log block from now on. */
+	if (srv_undo_log_encrypt) {
+		if (undo::spaces->empty()) {
+			srv_undo_log_encrypt = false;
+			ib::error() << cant_set_undo_tablespace << "s"
+				<< to_be_encrypted
+				<< ", since innodb_undo_tablespaces=0.";
+			return;
+		}
+
+		if (srv_read_only_mode) {
+			srv_undo_log_encrypt = false;
+			ib::error() << cant_set_undo_tablespace << "s"
+				<< to_be_encrypted
+				<< " in read-only mode.";
+			return;
+		}
+
+		undo::spaces->s_lock();
+		for (undo::Tablespaces_Vector::iterator undo_space_ptr =
+			undo::spaces->m_spaces.begin();
+			undo_space_ptr != undo::spaces->m_spaces.end();
+			++undo_space_ptr)
+		{
+			undo::Tablespace* undo_space = *undo_space_ptr;
+
+			/* Skip system tablespace, since it's also shared
+			tablespace. */
+			if (undo_space->id() == TRX_SYS_SPACE) {
+				continue;
+			}
+
+			space = fil_space_get(undo_space->id());
+			ut_ad(fsp_is_undo_tablespace(undo_space->id()));
+
+			ulint	new_flags =
+				space->flags | FSP_FLAGS_MASK_ENCRYPTION;
```

Comment:
I see you are setting FSP_FLAGS_MASK_ENCRYPTION here, but you are not saving it in undo tablespace header (you would need to update data dictionary for that too)  - which makes this flag temporary. What happens when we restart or crash - how the server will know that this tablespace is encrypted ?

Reply:
I do not think we can relay on MAGIC - as this can be correct by accident.